src/base/bittorrent/tracker.cpp

```diff
@@ -45,32 +47,153 @@ static const int ANNOUNCE_INTERVAL = 1800; // 30min
+            boost::asio::ip::address_v6 ipv6 = ip.to_v6();
+            auto addr = ipv6.to_bytes();
+            return ((addr[0] >> 1) == (0xFC >> 1))  // FC00::/7   Unique local
+                || ipv6.is_link_local()     // FE80::/10  Link-local
+                // ::ffff:0:0/96 ipv4-mapped of non-routable address
+                || (ipv6.is_v4_mapped() && isNonRoutable(ipv6.to_v4()))
+                // 2002::/16 6to4 ipv6 address
+                || (addr[0] == 0x20 && addr[1] == 0x02
+                    && isNonRoutable(toAddress(addr[2], addr[3], addr[4], addr[5])))
+                // 64:ff9b::/96
+                || (addr[0] == 0 && addr[1] == 0x64 && addr[2] == 0xFF && addr[3] == 0x9B
+                    && std::all_of(&addr[4], &addr[12], [](char a) { return a == 0; })
+                    && isNonRoutable(toAddress(addr[12], addr[13], addr[14], addr[15])));
+        }
+
+        return true;
+    }
+
+    // percent-escapes given string
+    std::string escape(const QByteArray& str)
+    {
+        // like libtorrent percent-encoding
+        QByteArray encoded = str.toPercentEncoding("!()*");
+        return std::string(encoded.data(), encoded.length());
+    }
 }
 
-bool Peer::operator==(const Peer &other) const
+// Peer
+struct Peer
 {
-    return uid() == other.uid();
-}
+    // ip retrieved from connection
+    address ip;
+    // ip obtained from announce query
+    address announceIp;
+    QByteArray peerId;
+    int port;
+    QByteArray uid;
+
+    Peer(const Peer&) = default;
+    Peer& operator=(const Peer&) = default;
+
+    // prevent move, as some boost libraries are compiled without move support 
+    Peer(Peer&&) = delete;
+    Peer& operator=(Peer&&) = delete;
+
+    // required by QMap
+    Peer() = default;
+    Peer(QString ipStr, QByteArray announceIpStr, int port, QByteArray peerId)
+        : port(port)
+        , peerId(peerId)
+    {
+        boost::system::error_code ec;
+        // prevent move, as some boost libraries are compiled without move support 
+        ip = static_cast<const address>(address::from_string(Utils::String::toStdString(ipStr), ec));
+        announceIp = static_cast<const address>(address::from_string(string(announceIpStr), ec));
+
+        uid = peerId.length() == 20 ? peerId : QString("%1:%2").arg(ipStr).arg(port).toLatin1();
+    }
+
+    bool operator!=(const Peer &other) const
+    {
+        return uid != other.uid;
+    }
+
+    bool operator==(const Peer &other) const
+    {
+        return uid == other.uid;
+    }
+
+    // if 'clientIP' belongs to a public network, returns public address found among connection and announce ip adrresses
+    const address& getIpForClient(const address& clientIp) const
+    {
+        // if any address is unspecified, return ip
+        if (clientIp.is_unspecified() || ip.is_unspecified() || announceIp.is_unspecified())
+            return ip;
+
+        // if client has public ip, return public ip
+        if (!isNonRoutable(clientIp)) {
```

Comment:
The purpose is to give external ip to external client and otherwise - connection ip. Otherwise two local clients on the same network behind a NAT router might not be able to share private torrents. If public ip is announced by default(inferred router's external ip and set in Session::handleExternalIPAlert()) and tracker returns external announceIp to local clients, then they will try to connect to the router and fail.